security/src/main/java/com/fincity/security/service/UserService.java

```diff
@@ -400,5 +415,212 @@ public Mono<Boolean> removingPermissionFromUser(ULong userId, ULong permissionId
+				                passwordPolicy)
+				        : Mono.empty(),
+
+		        (userId, isUpdatable, clientId, passwordPolicy, isValid,
+		                addPastPassword) -> addPastPassword.booleanValue()
+		                        ? this.dao.updatePassword(reqUserId, requestPassword.getNewPassword())
+		                        : Mono.empty(),
+
+		        (userId, isUpdatable, clientId, passwordPolicy, isValid, addPastPassword, passwordUpdated) ->
+				{
+
+			        if (passwordUpdated.booleanValue()) {
+				        this.soxLogService.create(new SoxLog().setObjectId(userId)
+				                .setActionName(SecuritySoxLogActionName.OTHER)
+				                .setObjectName(SecuritySoxLogObjectName.USER)
+				                .setDescription("Password updated"))
+				                .subscribe();
+
+				        this.dao.makeUserActive(reqUserId)
+				                .subscribe(); // making user active if it is another state
+
+				        return Mono.just(true);
+			        }
+
+			        return Mono.empty();
+
+		        }
+
+		).log()
+
+		        .switchIfEmpty(securityMessageResourceService.throwMessage(HttpStatus.FORBIDDEN,
+		                "Password cannot be updated"));
+
+	}
+
+	// add past password check here with dao and service here
+
+	public Mono<UShort> getHistoryCount(ULong clientId) {
+
+		return flatMapMono(
+
+		        () -> clientService.getClientPasswordPolicy(clientId),
+
+		        passwordPolicy -> passwordPolicy != null && passwordPolicy.getPassHistoryCount() != null
+		                ? Mono.just(passwordPolicy.getPassHistoryCount())
+		                : Mono.just(UShort.valueOf(0)));
+
+	}
+
+	// delete past password
+
+	public Mono<List<String>> fetchList(ULong userId) {
+		return this.dao.fetchPasswordsById(userId);
+	}
+
+	// add new password based on history count which was applicable for that
+	// selected user id
+
+	public Mono<Boolean> addBasedOnHistoryCount(ULong reqUserId, String password, ULong lUserId,
+	        ClientPasswordPolicy passwordPolicy) {
+
+		return flatMapMono(
+
+		        () -> Mono.just(reqUserId),
+
+		        userId -> this.dao.fetchPasswordsCount(userId),
+
+		        (userId, pastPasswordCount) -> passwordPolicy.getPassHistoryCount() != null
+		                ? Mono.just(passwordPolicy.getPassHistoryCount())
+		                : Mono.just(0),
+
+		        (userId, pastPasswordCount, historyCount) ->
+				{
+
+			        if (historyCount != null && pastPasswordCount >= historyCount.intValue() && pastPasswordCount > 0) {
+				        deletePastPassword(userId).subscribe();
+			        }
+
+			        this.dao.addPastPassword(reqUserId, password, lUserId)
+			                .subscribe();
+			        return Mono.just(true);
+
+		        }
+
+		).log();
+
+	}
+
+	// delete past password record based on oldest time stamp for selected userId
+	public Mono<Boolean> deletePastPassword(ULong userId) {
```

Comment:
You should not have a delete past password at all.

Reply:
removed this method